src/nvim/eval/decode.c

```diff
@@ -0,0 +1,1116 @@
+  }
+  if (*next_map_special) {
+    goto parse_json_number_ret;
+  }
+  p--;
+  goto parse_json_number_ret;
+parse_json_number_fail:
+  ret = FAIL;
+parse_json_number_ret:
+  *pp = p;
+  return ret;
+}
+
+#define POP(obj_tv, is_sp_string) \
+  do { \
+    if (json_decoder_pop(OBJ(obj_tv, is_sp_string, didcomma, didcolon), \
+                         &stack, &container_stack, \
+                         &p, &next_map_special, &didcomma, &didcolon) \
+        == FAIL) { \
+      goto json_decode_string_fail; \
+    } \
+    if (next_map_special) { \
+      goto json_decode_string_cycle_start; \
+    } \
+  } while (0)
+
+/// Convert JSON string into VimL object
+///
+/// @param[in]  buf  String to convert. UTF-8 encoding is assumed.
+/// @param[in]  buf_len  Length of the string.
+/// @param[out]  rettv  Location where to save results.
+///
+/// @return OK in case of success, FAIL otherwise.
+int json_decode_string(const char *const buf, const size_t buf_len,
+                       typval_T *const rettv)
+  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT
+{
+  const char *p = buf;
+  const char *const e = buf + buf_len;
+  while (p < e && (*p == ' ' || *p == TAB || *p == NL || *p == CAR)) {
+    p++;
+  }
+  if (p == e) {
+    EMSG(_("E474: Attempt to decode a blank string"));
+    return FAIL;
+  }
+  vimconv_T conv = { .vc_type = CONV_NONE };
+  convert_setup(&conv, (char_u *) "utf-8", p_enc);
+  conv.vc_fail = true;
+  int ret = OK;
+  ValuesStack stack;
+  kv_init(stack);
+  ContainerStack container_stack;
+  kv_init(container_stack);
+  rettv->v_type = VAR_UNKNOWN;
+  bool didcomma = false;
+  bool didcolon = false;
+  bool next_map_special = false;
+  for (; p < e; p++) {
+json_decode_string_cycle_start:
+    assert(*p == '{' || next_map_special == false);
+    switch (*p) {
+      case '}':
+      case ']': {
+        if (kv_size(container_stack) == 0) {
+          emsgf(_("E474: No container to close: %.*s"), LENP(p, e));
+          goto json_decode_string_fail;
+        }
+        ContainerStackItem last_container = kv_last(container_stack);
+        if (*p == '}' && last_container.container.v_type != VAR_DICT) {
+          emsgf(_("E474: Closing list with curly bracket: %.*s"), LENP(p, e));
+          goto json_decode_string_fail;
+        } else if (*p == ']' && last_container.container.v_type != VAR_LIST) {
+          emsgf(_("E474: Closing dictionary with square bracket: %.*s"),
+                LENP(p, e));
+          goto json_decode_string_fail;
+        } else if (didcomma) {
+          emsgf(_("E474: Trailing comma: %.*s"), LENP(p, e));
+          goto json_decode_string_fail;
+        } else if (didcolon) {
+          emsgf(_("E474: Expected value after colon: %.*s"), LENP(p, e));
+          goto json_decode_string_fail;
+        } else if (last_container.stack_index != kv_size(stack) - 1) {
+          assert(last_container.stack_index < kv_size(stack) - 1);
+          emsgf(_("E474: Expected value: %.*s"), LENP(p, e));
```

Comment:
We expect both `:` and value. Value is more significant: e.g. in Python `{"a", "b"}` is a valid set literal.